### PR TITLE
[OCPCLOUD-1354] Allow Machines to select EFA network interfaces

### PR DESCRIPTION
**Pull request by JoelSpeed**
This implements the Machine API part of https://issues.redhat.com/browse/RFE-2236 by using the new fields introduced in openshift/api#1065

I have manually tested this by creating a machineset with a `c6i.32xlarge` instance in the us-east-2 region with `networkInterfaceType: efa`. The instance joined the cluster successfully and became a fully running node.

We must wait until the API PR merges before we can merge this.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -61,6 +61,8 @@ github.com/Azure/go-autorest/tracing v0.5.0/go.mod h1:r/s2XiOKccPW3HrqB+W0TQzfbt
 github.com/Azure/go-autorest/tracing v0.6.0/go.mod h1:+vhtPC754Xsa23ID7GlGsrdKBpUA79WCAKPPZVC2DeU=
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym/WlBOVXweHU+Q+/VP0lqqI8lqeDx9IjBqo=
+github.com/JoelSpeed/api v0.0.0-20211123141226-cbad8a97a6a6 h1:EA40WvSPBjG2msBgikyr/1YFDU2GKsAoSME3d/QGdEc=
+github.com/JoelSpeed/api v0.0.0-20211123141226-cbad8a97a6a6/go.mod h1:RsQCVJu4qhUawxxDP7pGlwU3IA4F01wYm3qKEu29Su8=
```
Comment (elmiko) [inline go.sum:65]: don't forget to change these back ;)

**Review by elmiko — COMMENTED** [review]
this makes sense to me, just left a comment to remember the imports

File: pkg/actuators/machine/instances.go [hunk 2]
```diff
@@ -310,6 +310,17 @@ func launchInstance(machine *machinev1.Machine, machineProviderConfig *machinev1
 		},
 	}
 
+	switch machineProviderConfig.NetworkInterfaceType {
+	case machinev1.AWSENANetworkInterfaceType:
+		networkInterfaces[0].InterfaceType = aws.String("interface")
+	case machinev1.AWSEFANetworkInterfaceType:
+		networkInterfaces[0].InterfaceType = aws.String("efa")
+	case "":
+		// Do nothing
```
Comment (lobziik) [inline pkg/actuators/machine/instances.go:319]: I'm not quite understand, why nothing here.
Would it make sense to put reasonable default there, just for make this behaviour clear (`aws.String("interface")` maybe).

File: pkg/actuators/machine/instances_test.go [hunk 3]
```diff
@@ -886,6 +886,85 @@ func TestLaunchInstance(t *testing.T) {
...
+							aws.String("sg-009a70e28fa4ba84e"),
+							aws.String("sg-07323d56fb932c84c"),
+							aws.String("sg-08b1ffd32874d59a2"),
+						},
+						InterfaceType: aws.String("efa"),
+					},
+				},
+				UserData: aws.String(""),
+			},
+		},
```
Comment (lobziik) [inline pkg/actuators/machine/instances_test.go:967]: Would be nice to have testcase for empty `NetworkInterfaceType`

File: pkg/actuators/machine/instances.go (hunk 2 above)
Comment (alexander-demichev) [reply pkg/actuators/machine/instances.go:319]: networkInterfaces[0].InterfaceType is pointer, I guess that default is nil

**Review by alexander-demichev — COMMENTED** [review]
/approve

File: pkg/actuators/machine/instances.go (hunk 2 above)
Comment (lobziik) [reply pkg/actuators/machine/instances.go:319]: Or just more detailed comment

Comment (lobziik) [reply pkg/actuators/machine/instances.go:319]: >networkInterfaces[0].InterfaceType is pointer, I guess that default is nil

File: pkg/actuators/machine/instances_test.go (hunk 3 above)
Comment (JoelSpeed) [reply pkg/actuators/machine/instances_test.go:967]: All of the other test cases have an empty `NetworkInterfaceType` and so they cover the case where it isn't set already

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-aws-upgrade

The upgrades are failing across the board at the moment, I'm pretty confident the failures aren't caused by this PR

**Pull request merged**